Lib/typing.py

```diff
@@ -1060,15 +1060,25 @@ def _evaluate(self, globalns, localns, type_params=_sentinel, *, recursive_guard
                 globalns = getattr(
                     sys.modules.get(self.__forward_module__, None), '__dict__', globalns
                 )
-            if type_params:
-                # "Inject" type parameters into the local namespace
-                # (unless they are shadowed by assignments *in* the local namespace),
-                # as a way of emulating annotation scopes when calling `eval()`
-                locals_to_pass = {param.__name__: param for param in type_params} | localns
+
+            # type parameters require some special handling,
+            # as they exist in their own scope
+            # but `eval()` does not have a dedicated parameter for that scope.
+            # For classes, names in type parameter scopes should override
+            # names in the global scope (which here are called `localns`!),
+            # but should in turn be overridden by names in the class scope
+            # (which here are called `globalns`!)
+            type_params = {param.__name__: param for param in type_params}
+            if self.__forward_is_class__:
+                for name, param in type_params.items():
+                    if name not in globalns:
+                        globalns[name] = param
```

Comment:
I don't think we can safely mutate the globalns. Note that in some circumstances globalns is a module dictionary.

Reply:
Thanks, fixed in https://github.com/python/cpython/pull/120272/commits/5505d971a667cdcc10c31ace527f411e22112213